zkevm-circuits/src/evm_circuit/op_execution/comparator.rs

```diff
@@ -1,379 +1,205 @@
-                name: "LtGadget success",
-                selector: selector.expr(),
-                polys: [
-                    state_transition_constraints,
-                    swap_constraints,
-                    sumc_constraints,
-                    sumc_is_zero_constraints,
-                    lt_constraints,
-                    result_constraints,
-                ]
-                .concat(),
-                lookups: [c_range_lookups, bus_mapping_lookups].concat(),
-            }
-        };
+        name: &'static str,
+    ) -> Constraint<F> {
+        let mut cb = ConstraintBuilder::default();
+
+        // Check if this is EQ
+        let is_eq_op = self.is_eq_op.constraints(
+            &mut cb,
+            state_curr.opcode.expr(),
+            OpcodeId::EQ.expr(),
+        );
 
-        let stack_underflow = {
-            let stack_pointer = state_curr.stack_pointer.expr();
-            Constraint {
-                name: "LtGadget stack underflow",
-                selector: self.stack_underflow.expr(),
-                polys: vec![
-                    (stack_pointer.clone() - 1024.expr())
-                        * (stack_pointer - 1023.expr()),
-                ],
-                lookups: vec![],
-            }
-        };
+        // For GT we swap the stack inputs so that we actually do
+        // greater than instead of smaller than.
+        let swap = self.swap.constraints(
+            &mut cb,
+            state_curr.opcode.expr(),
+            OpcodeId::GT.expr(),
+        );
 
-        let out_of_gas = {
-            let (selector, gas_available) = &self.out_of_gas;
-            let gas_overdemand = state_curr.gas_counter.expr()
-                + GasCost::FASTEST.expr()
-                - gas_available.expr();
-            Constraint {
-                name: "LtGadget out of gas",
-                selector: selector.expr(),
-                polys: vec![
-                    (gas_overdemand.clone() - 1.expr())
-                        * (gas_overdemand.clone() - 2.expr())
-                        * (gas_overdemand - 3.expr()),
-                ],
-                lookups: vec![],
-            }
-        };
+        // `a[0..16] <= b[0..16]`
+        let (lt_lo, eq_lo) = self.comparison_lo.constraints(
+            &mut cb,
+            from_bytes::expr(self.a.cells[0..16].to_vec()),
+            from_bytes::expr(self.b.cells[0..16].to_vec()),
+        );
 
-        array::IntoIter::new([success, stack_underflow, out_of_gas])
-            .map(move |mut constraint| {
-                constraint.polys =
-                    [common_polys.clone(), constraint.polys].concat();
-                constraint
-            })
-            .collect()
-    }
+        // `a[16..32] <= b[16..32]`
+        let (lt_hi, eq_hi) = self.comparison_hi.constraints(
+            &mut cb,
+            from_bytes::expr(self.a.cells[16..32].to_vec()),
+            from_bytes::expr(self.b.cells[16..32].to_vec()),
+        );
```

Comment:
> That's an interesting way of looking at it! I think this would definitely work if we didn't have to do the comparison in two parts. But because of the two separate parts I don't think there's a 1:1 relation to the `swap` value for both parts (the first part can be smaller while the second part can be bigger, so the equations may need to be made valid in different ways). I do think there's probably a way to use the `swap` value directly in just the `hi` part because some 1:1 mapping could probably be found there... But I would guess (and very much a guess!) this probably wouldn't make the efficiency of the circuit better (in the best case save a single cell with maybe some more complex expressions in return because `swap` isn't just `1` when GT) and the current method allows the reuse of a general gadget that can be used everywhere which is nice. Am I still missing something or would you agree?

I like the implementation and abstraction, but I am wondering whether some constraint is missing to somehow connect swap and lt? For example, if you change

`check(OpcodeId::LT, hi_lo.clone(), lo_hi.clone(), 0x00u64.into());
`
to
`check(OpcodeId::GT, hi_lo.clone(), lo_hi.clone(), 0x00u64.into());
`

and if you change assign function (which prover can do) from
`let swap = step.opcode == OpcodeId::GT;
`
to
`let swap = false;`

the test will pass.

Reply:
That's because currently the compressed value of both `hi_lo` and `lo_hi` is the same because `r` is set to `1`: https://github.com/appliedzkp/zkevm-circuits/blob/dfff7bd1d46098e7c3c1d51929cb904d3ac0b662/zkevm-circuits/src/evm_circuit.rs#L981-L982

So the the compression simply does a sum of all byte values, which for `hi_lo` and `lo_hi` is the same. If you use values where this isn't the case that will fail as expected. Also note that the `swap` value that was changed in your example only impacts the swap of the stack values, not the constraint that checks what the opcode actually is (the witness for that is calculated in the `IsEqualGadget`). So definitely a bit confusing because of these things together why that worked :).

I think all tests currently depend on `r == 1` for generating the compressed values for the test values because `r` is not directly made available as far as I know.

Reply:
Thanks! Didn't know about `r = 1`. I changed it to 2, but then all tests fail (unchanged) - do I need to change something else to enable different `r`?

Yes, I am aware of `swap` affecting only the stack order - that's actually where my concern came from, because it seemed to me that we can for example provide proof for a < b, but then we use the same proof for a > b by switching a and b in the proof, because the information about being LT or GT is on one hand in the swap and on the other hand in opcode/result, but somehow it seems to me that this information is not checked to be the same. Probably I am just missing something. I am sorry for taking your time, I am probably just a bit too paranoid about forgetting some constraints.

Reply:
> Thanks! Didn't know about r = 1. I changed it to 2, but then all tests fail (unchanged) - do I need to change something else to enable different r?

Modified the `compress` helper function used in the comparator tests to use a hardcoded `r` value. If you change it to match the one set on the circuit it should work. Currently hard coded so will need to be improved in the future so we can get this value from withing `assign` somehow. A lot of tests currently explicitly sum the bytes manually for the compression so definitely some general test improvements necessary. I left the test cases intact in my other refactor PR because those PRs already change enough, but something that needs to be done in a future PR.

> Yes, I am aware of `swap` affecting only the stack order - that's actually where my concern came from, because it seemed to me that we can for example provide proof for a < b, but then we use the same proof for a > b by switching a and b in the proof, because the information about being LT or GT is on one hand in the swap and on the other hand in opcode/result, but somehow it seems to me that this information is not checked to be the same. Probably I am just missing something. I am sorry for taking your time, I am probably just a bit too paranoid about forgetting some constraints.

The direct connection is there for the constraints right? In the witness generation there were indeed two independent variables, but that's just for setting the witnesses so it doesn't matter. But I changed the witness assignment to more closely follow how the constraints work, I hope that makes things more clear how everything is connected with the constraints.

Reply:
Thanks for your patience, Brecht! Sorry for troubling you!